htdocs/js/PGCodeMirror/PG.js

```diff
@@ -1418,16 +1418,33 @@
 			lineComment: '#'
 		};
 	});
-	
+
 	CodeMirror.registerHelper("wordChars", "perl", /[\w$]/);
-	
+
+	CodeMirror.registerHelper('fold','PG', (cm,start) => {
+		const m1 = /BEGIN_(PGML|PGML_SOLUTION|PGML_HINT|TEXT|TIKZ|LATEX_IMAGE)\s*$/.exec(cm.getLine(start.line));
+		const m2 = /^\s*(Section|Scaffold)::Begin/.exec(cm.getLine(start.line));
+		if(m1 || m2 ){
+			for (let current_line = start.line +1; current_line <= cm.lineCount(); current_line++){
+				const end_re = m1 ? RegExp(`END_${m1[1]}`) : RegExp(`${m2[1]}::End`);
+				if(end_re.test(cm.getLine(current_line))) {
+					return {
+						from: CodeMirror.Pos(start.line, cm.getLine(start.line).length),
+						to: CodeMirror.Pos(current_line, cm.getLine(current_line).length)
+					};
+				}
+			}
+		}
+		return;
+	});
```

Comment:
I didn't look closely at this before.  The regular expressions need to be more precise.  If a line contains `# BEGIN_PGML` it should not start a fold.  Also, there is no need for the m1 and m2 variables.  Here is my suggested way to do this:
```suggestion
	CodeMirror.registerHelper('fold', 'PG', (cm, start) => {
		const match =
			/^\s*BEGIN_(PGML|PGML_SOLUTION|PGML_HINT|TEXT|TIKZ|LATEX_IMAGE)\s*$/.exec(cm.getLine(start.line)) ||
			/^\s*(Section|Scaffold)::Begin/.exec(cm.getLine(start.line)) ||
			/^[$\w]*->\s*BEGIN_(TIKZ|LATEX_IMAGE)\s*$/.exec(cm.getLine(start.line));
		if (match) {
			for (let current_line = start.line + 1; current_line <= cm.lineCount(); ++current_line) {
				if (RegExp(`END_${match[1]}`).test(cm.getLine(current_line))) {
					return {
						from: CodeMirror.Pos(start.line, cm.getLine(start.line).length),
						to: CodeMirror.Pos(current_line, cm.getLine(current_line).length)
					};
				}
			}
		}
		return;
	});
```

Reply:
Sorry.  You do need the `m1` and `m2`.  I missed the difference between the scaffolds and the others.  So make that:
```suggestion
	CodeMirror.registerHelper('fold', 'PG', (cm, start) => {
		const m1 =
			/^\s*BEGIN_(PGML|PGML_SOLUTION|PGML_HINT|TEXT|TIKZ|LATEX_IMAGE)\s*$/.exec(cm.getLine(start.line)) ||
			/^[$\w]*->\s*BEGIN_(TIKZ|LATEX_IMAGE)\s*$/.exec(cm.getLine(start.line));
		const m2 = /^\s*(Section|Scaffold)::Begin/.exec(cm.getLine(start.line));
		if (m1 || m2) {
			for (let current_line = start.line + 1; current_line <= cm.lineCount(); ++current_line) {
				const end_re = m1 ? RegExp(`END_${m1[1]}`) : RegExp(`${m2[1]}::End`);
				if(end_re.test(cm.getLine(current_line))) {
					return {
						from: CodeMirror.Pos(start.line, cm.getLine(start.line).length),
						to: CodeMirror.Pos(current_line, cm.getLine(current_line).length)
					};
				}
			}
		}
		return;
	});

Reply:
The regular expressions in my suggestion more closely match the perl regular expressions in the translator.

Reply:
I removed the `|TIKZ|LATEX_IMAGE` from the regex on line 1443 since it is on 1444.

Reply:
Yes, it should have been removed.  It was in my initial edits, but then it was undone later in successive changes.